### DISCUSSION TITLE
Is AST necessary?

### DISCUSSION THREAD

**Discussion by yisar** [biomejs/biome#703] [Q&A] [answered]
I have read Boime's article and noticed that there are two structures here, CST and AST, both of which can be traversed and modified. So why not directly traverse CST? AST is a higher level abstraction, which is not a good choice for Linter

**Comment by ematipico** [comment]
I don't know which article you're talking about, but at this page you can find more details

https://biomejs.dev/internals/architecture/#parser-and-cst

Biome doesn't have any AST, the playground is misleading and it's just a way to ease new users that aren't familiar with the concept.

**Reply by yisar** [reply]
Okay, I was misled. I don't understand why AST structures are needed in playground, and there are also many ASTNode structures in the source code, which is indeed a misleading approach.

**Reply by ematipico** [reply, marked as answer]
The box that shows the AST is what you actually query/navigate when you want to do operations like lint rules or formatting logic.

It becomes an indispensable tool when developing features and fixing bugs.

The CST box is the "raw" language, the AST is strongly typed and shows the tokens/nodes of the syntax APIs
